gem/gem.py

```diff
@@ -782,6 +783,58 @@ def __new__(cls, i, j):
         return self
 
 
+class Inverse(Node):
+    """ Inverse is a shape to shape operation.
+
+        :arg aggregate: Tensor to be inversed
+        :arg multiindex: Indices for result of Inverse
+
+        Inverse is different from ComponentTensor, because its children must have shape.
+        It does no transformation between free indices and shape, but the multiindex is still necessary.
+        The multiindices of the children are used for indexing them when passing to inverse function call,
+        the multiindex of the Inverse node itself is used to index the result of the function call.
+    """
+    __slots__ = ('children', 'shape')
+    __back__ = ()
+
+    def __new__(cls, aggregate):
+        assert aggregate.shape
+
+        self = super(Inverse, cls).__new__(cls)
+        self.children = (aggregate,)
+        self.shape = aggregate.shape
+        self.free_indices = aggregate.free_indices
+
+        return self
+
+
+class Solve(Node):
+    """ Solve is a shape to shape operation similar to Inverse.
+
+        :arg A: Tensor
+        :arg B: Right-hand side
+        :arg multiindex: Indices for result of solve
+    """
+    __slots__ = ('children', 'shape')
+    __back__ = ()
+
+    def __new__(cls, A, B):
+
+        # Children must have shape
+        assert A.shape
+        assert B.shape
+
+        # Match shapes
+        assert A.shape[1] == B.shape[0]
+
+        self = super(Solve, cls).__new__(cls)
+        self.children = (A, B)
+        self.shape = B.shape
+        self.free_indices = B.free_indices
```

Comment:
Yes, this is not right indeed.
Your `A` should have two indices, no? And this is a contraction in the end, which is why we need to guarantee the shape matching you mentioned above, so we would also need to "loose" an index, right?
Maybe I should something like `self.free_indices = unique(set(A.free_indices) - set(B.free_indices))`?

Reply:
Well, I thought we had decided that `Solve` operates on shape, so it represents `A^{-1} B` (and therefore the free indices of the output are the union of the free indices of the input). There's no contraction on the free indices of A and B, I think.

We have `A.shape = (N, N)` with some free indices, and `B.shape = (N, O, P, Q, ...)` with some free indices. So then the matching of shapes has to be (since inverting A flips the shape axes) `A.shape[0] == B.shape[0]` and so we end up with something of shape `(N, O, P, Q, ...)` but it still has free indices merging those of A and B.

Reply:
Ah right, yes.